04-store-api/starter/controllers/products.js

```diff
@@ -0,0 +1,75 @@
+const Product = require("../models/product");
+
+const getAllProductsStatic = async (req, res) => {
+  const products = await Product.find({ price: { $gt: 30 } })
+    .sort("price")
+    .select("name price");
+  res.status(200).json({ products, nbHits: products.length });
+};
+
+const getAllProducts = async (req, res) => {
+  const { featured, company, name, sort, fields, numericFilters } = req.query;
+  const queryObject = {};
+  if (featured) {
+    queryObject.featured = featured === "true" ? true : false;
+  }
+  if (company) {
+    queryObject.company = company;
+  }
+  if (name) {
+    queryObject.name = { $regex: name, $options: "i" };
```

Comment:
Research whether there are security implications for passing user input to mongodb as a regex. I am a little suspicious, but maybe mongo's API is hardened to the extent that this is OK.

There is a class of attacks called Denial of Service (DoS) attacks, where a hacker creates an input that locks up your system because it takes a very long time to return a response. For example, if you can create an input that creates an infinite loop, it will lock up the server that receives that request. This is a big deal in enterprise systems, because if you can send an request basically instantly and lock up a server for a few seconds, you can just hammer the system with those costly requests and completely deny the requests from legitimate clients since the system is so overwhelmed.

Regexes can be crafted in a way that they are massively expensive to evaluate or even to compile, so I am suspicious about this usage.

Reply:
Sidenote: maybe you can use `$search` instead; seems like it matches the use case better https://www.mongodb.com/docs/atlas/atlas-search/query-syntax/#mongodb-pipeline-pipe.-search

Also, I was trying out the regex support and I'm not sure if it was working right; although I don't really know mongodb's specific regex syntax. For example, I tried this URL:

http://localhost:3000/api/v1/products?price>30?name=%5Em

If you url decode the "%5" it translates to "^", so I'm trying to use the regex `^m` (any name starting with `m`). It didn't seem to have any impact on the search results though.

Reply:
TODO: edit prev comment; that's a malformed URL. Using `$search` still seems advisable.